File: bot/commands/reportCarryOverCommand.py

from discord.raw_models import RawReactionActionEvent
from apiClient import ApiClient
from command_spec import PhraseRepository, ReactionCommand


class ReportCarryOverCommand(ReactionCommand):
    def __init__(self, phraseRepository: PhraseRepository, apiClient: ApiClient) -> None:
        super().__init__(phraseRepository)
        self.apiClient = apiClient

    def isMatchTo(self, reaction: str, messageId: str) -> bool:
        response = self.apiClient.get_report_message(messageId)

        if(response.status_code == 200):
            return (
                reaction == self.phraseRepository.get('first_carry_over_stamp')
                or reaction == self.phraseRepository.get('second_carry_over_stamp')
            )
        else:
            return False

    def executeForAdd(self, reaction: RawReactionActionEvent) -> None:
        print("report carry over")
        self.apiClient.report_carry_over(reaction.message_id, reaction.user_id)

    def executeForRemove(self, reaction: RawReactionActionEvent) -> None:
        print("cancel carry over")
        self.apiClient.cancel_carry_over(reaction.message_id, reaction.user_id)

File: bot-js/src/commands/ReportCarryOverCommand.ts
import { MessageReaction, User } from "discord.js";
import { ApiClient } from "../backend/ApiClient";
import { PhraseKey } from "../support/PhraseKey";
import { PhraseRepository } from "../support/PhraseRepository";
import { ReactionCommand } from "./ReactionCommand";

export class ReportCarryOverCommand implements ReactionCommand {
    constructor(private phraseRepository: PhraseRepository, private apiClient: ApiClient) {}

    async isTarget(reaction: MessageReaction): Promise<boolean> {
        return (
            (await this.apiClient.hasReportMessage(reaction.message.id)) &&
            (reaction.emoji.toString() === this.phraseRepository.get(PhraseKey.firstCarryOverStamp()) ||
                reaction.emoji.toString() === this.phraseRepository.get(PhraseKey.secondCarryOverStamp()))
        );
    }

    async executeForAdd(reaction: MessageReaction, user: User): Promise<void> {
        if (!(await this.isTarget(reaction))) return;
        console.log("report carry over");

        await this.apiClient.reportCarryOver(reaction.message.id, user.id);
    }

    async executeForRemove(reaction: MessageReaction, user: User): Promise<void> {
        if (!(await this.isTarget(reaction))) return;
        console.log("cancel carry over");

        await this.apiClient.cancelCarryOver(reaction.message.id, user.id);
    }
}
